### DISCUSSION TITLE
NeighborSampler vs. NeighborLoader

### DISCUSSION THREAD

**Discussion by pintonos** [pyg-team/pytorch_geometric#3699] [General]
Hello!
What is the difference between the NeighborSampler and the NeighborLoader? I am currently working in an inductive setting and the only example in the repo uses NeighborSampling in combination with SageConv. However, for my specific use case (additional edge weights + GCNConv) generating batches via the NeighborLoader would be more convenient. So I was wondering if they differ in the inductive case?

Thanks for the help!

**Comment by rusty1s** [comment]
We recommend the usage of `NeighborLoader`, which streamlines the interface of `NeighborSampler`: `NeighborLoader` will just return a single `data` object (not a list of bipartite graphs), which makes it much more broadly applicable and easier-to-use. The functionality/output embeddings should be the same in either way :)

**Reply by pintonos** [reply]
OK, thanks for the info!

To ensure correctness:
```
train_node_idx = torch.where(data.train_mask)[0]
train_loader = NeighborLoader(data, num_neighbors=[10, 10], shuffle=True,
                                  input_nodes=train_node_idx, batch_size=512)
```
, where `data` is the full dataset.

This usage is correct for the inductive training case?

**Reply by rusty1s** [reply]
No, it will not guarantee that you only sample training nodes. To apply it in an inductive learning scenario, you first have to filter out non-training nodes and only push a "subgraph" data object only holding training nodes to the loader:
```python
train_loader = NeighborLoader(data.subgraph(data.train_mask), num_neighbors=[10, 10], shuffle=True,
                              input_nodes=None, batch_size=512)
```
Now, `input_nodes` can be `None` (it will sample neighbors for all nodes in the graph). Note that `data.subgraph(...)` is a rather new method (you may need to install from master or manually filter out your `data` object via `torch_geometric.utils.subgraph`).

**Reply by pintonos** [reply]
This code seems to work fine on CPU, but gets a Segmentation Fault on the GPU:
```
train_edge_index, train_edge_weight = subgraph(data.train_mask, edge_index=data.edge_index, edge_attr=data.edge_weight)
train_graph = Data(data.x, train_edge_index, edge_weight=train_edge_weight)
train_loader = NeighborLoader(train_graph, num_neighbors=[10, 10], shuffle=True,
                              input_nodes=None, batch_size=512)
```
Any idea why?

**Reply by rusty1s** [reply]
`NeighborLoader` can only process CPU tensors, it does not have a GPU implementation. As such, this error is to be expected (although it might be great to add a more meaningful error message).

**Reply by AntonCoon** [reply]
Is it correct to understand that, for GPU usage, NeighborSampler is the sole option without any alternatives?

**Reply by EdisonLeeeee** [reply]
No, `NeighborSampler` also does not have GPU support.
